clean_build.sh

```diff
@@ -1,32 +1,47 @@
 #!/bin/sh
+
 cd "$(dirname "$0")" || exit 1
+
 # some environments have cmake v2 as 'cmake' and v3 as 'cmake3'
 # check for cmake3 first then fallback to just cmake
-B_CMAKE=`type cmake3 2>/dev/null`
-if [ $? -eq 0 ]; then
-    B_CMAKE=`echo "$B_CMAKE" | cut -d' ' -f3`
+
+B_CMAKE=$(command -v cmake3 2>/dev/null)
+if [ "$?" -eq 0 ]; then
+    continue
 else
-    B_CMAKE=cmake
+    if command -v cmake 2>/dev/null; then
+        B_CMAKE=$(command -v cmake)
+    else
+        echo "ERROR: CMake not in $PATH, cannot build! Please install CMake, or if this persists, file a bug report."
+        exit 1
+    fi
 fi
-# default build configuration
-B_BUILD_TYPE=${B_BUILD_TYPE:-Debug}
+
+B_BUILD_TYPE="${B_BUILD_TYPE:-Debug}"
+B_CMAKE_FLAGS="-DCMAKE_BUILD_TYPE=${B_BUILD_TYPE} ${B_CMAKE_FLAGS:-}"
+
 if [ "$(uname)" = "Darwin" ]; then
-    # OSX needs a lot of extra help, poor thing
-    # run the osx_environment.sh script to fix paths
-    . ./osx_environment.sh
-    B_CMAKE_FLAGS="-DCMAKE_OSX_SYSROOT=$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9 $B_CMAKE_FLAGS"
+    # macOS needs a little help, so we source this environment script to fix paths.
+    . ./macos_environment.sh
+
+    B_CMAKE_FLAGS="${B_CMAKE_FLAGS} -DCMAKE_OSX_SYSROOT=$(xcode-select --print-path)/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -DCMAKE_OSX_DEPLOYMENT_TARGET=10.9"
 fi
-# allow local customizations to build environment
-[ -r ./build_env.sh ] && . ./build_env.sh
+
+# Source local build variables to the environment, if available.
+# If not, continue as normal, and silently.
+[ -e "./build_env.sh" ] && . "./build_env.sh"
```

Comment:
I think the previous version was better because we actually check whether the file exists and read permission is granted. Also no need to put quotes when arguments are not variables - they won't accidentally emit a space character.

Reply:
Yes, `-r` is fine too. I disagree that we shouldn't put quotes around arguments though. A fair amount of bash/sh scripts do put arguments in variables - in this case, I'd rather keep the quote around the conditional's arguments, and not use arguments in the sourcing if that works. The reason I kept it is for readability to be honest - I've looked at `acme.sh` for example, and it's quite a common practice there...

Reply:
Fair enough, but if we change this, then it's best to do in a single commit and change all occurrences of non-quoted paths in the file.

Reply:
OK, so I've changed all paths like that to be quoted. The `if -v command -v cmake 2>/dev/null` could be presented into a `$()` for example, would you want that? Otherwise, it wouldn't be quoted...? Same with the first assignment of `B_CMAKE`.

The other thing I'm cautious to commit so far is the `rm -rf build` changes. All there of those lines, `rm`, `mkdir`, and `cd` now have `"build"` quoted. Is that the behaviour you'd like to see?